File: src/common/common/main/java/settingdust/preloading_tricks/util/LoaderAdapter.java
package settingdust.preloading_tricks.util;

import com.google.common.base.Supplier;
import com.google.common.base.Suppliers;

public interface LoaderAdapter {
    Supplier<LoaderAdapter> INSTANCE = Suppliers.memoize(() -> ServiceLoaderUtil.findService(LoaderAdapter.class));

    static LoaderAdapter getInstance() {
        return INSTANCE.get();
    }

    boolean isClient();

    boolean isModLoaded(String modId);
}

File: src/fabric/common/main/kotlin/settingdust/preloading_tricks/fabric/util/LoaderAdapter.kt
package settingdust.preloading_tricks.fabric.util

import net.fabricmc.api.EnvType
import net.fabricmc.loader.api.FabricLoader
import settingdust.preloading_tricks.util.LoaderAdapter

class LoaderAdapter : LoaderAdapter {
    override val isClient = FabricLoader.getInstance().environmentType === EnvType.CLIENT

    override fun isModLoaded(modId: String) = FabricLoader.getInstance().isModLoaded(modId)
}